File: src/main/java/editor/collection/export/CardListFormat.java
package editor.collection.export;

import java.io.IOException;
import java.io.InputStream;
import java.text.ParseException;
import java.util.List;

import editor.collection.CardList;

/**
 * This is an interface for formatting a list of cards.  It can format
 * a list of cards into a string and it can parse a string into a card
 * list.
 *
 * @author Alec Roelke
 */
public interface CardListFormat
{
    /**
     * Create a string representation of the given list of cards.
     *
     * @param list list of cards to format
     * @return the string representation of the list of cards
     */
    String format(CardList list);

    /**
     * @return <code>true</code> if this CardListFormat has headings, and
     * <code>false</code> otherwise.
     */
    default boolean hasHeader()
    {
        return !header().isEmpty();
    }

    /**
     * @return A String containing the formatted list header.  If the format
     * doesn't contain headers, it should be the empty String.
     */
    String header();

    /**
     * Parse a string for a list of cards.
     *
     * @param source string to parse, split by lines
     * @return list of cards containing values from the parsed string
     * @throws ParseException if the string cannot be parsed
     */
    CardList parse(List<String> source) throws ParseException;

    CardList parse(InputStream source) throws ParseException, IOException;
}

File: src/main/scala/editor/collection/export/CardListFormat.scala
package editor.collection.`export`

import editor.collection.CardList
import editor.gui.deck.DesignSerializer

import java.io.InputStream

/**
 * Generic formatter for converting a [[CardList]] into a string or parsing a file into a deck.
 * @author Alec Roelke
 */
trait CardListFormat {
  /**
   * Convert a [[CardList]] into a string. Which cards, and which attributes of and other information about the cards,
   * that are included and how that information is arranged is implementation-dependent.
   * 
   * @param list list to convert
   * @return the formatted string representing the card list
   */
  def format(list: CardList): String

  /** The header of the string version of the list, or the None string if there isn't one. */
  def header: Option[String]

  /**
   * Parse a stream into a new deck. The format of the file is implementation-dependent.
   * 
   * @param source stream to convert to a deck
   * @return a [[DeckSerializer]] containing all of the information about the deck
   */
  def parse(source: InputStream): DesignSerializer
}
